### PR TITLE
Unify and clarify config loading

### PR DESCRIPTION
**Pull request by gibbz00**
The PR aims to reduce the omissions of necessary reliability checks when loading the configs (editor, theme, lang), whilst hopefully also improving code maintainability.  All in all, `xtask`, `helix_loader` and `helix_term::{main, appliction}` have been cleaned up quite significantly. Steps are documented in the commit messages. 

Relies on #5635

### Identify and merge PRs that clash with this one.
- [x] https://github.com/helix-editor/helix/pull/5656 
-  [x] https://github.com/helix-editor/helix/pull/5411
- [x] https://github.com/helix-editor/helix/pull/3207

My only gripe from #5411 was that it now places `hx --grammar (fetch|build)` files into `XDG_CONFIG_HOME` (`.config/`), under the condition that it exists. Would much rather prefer it to look in `XDG_DATA_HOME` (`.local/share`) first. Which is something I've added in this PR. This might be a solution to https://github.com/helix-editor/helix/issues/584, and an alternative to https://github.com/helix-editor/helix/pull/5603. In fact, #3202 is the improvement suggested by #584's author:

> Even better solution would include Helix merging config and grammars from system dirs and user dirs and having fallbacks configurable at buildtime but meh. This should be enough for most usecases.

Does also not change log file location to an OS state directory, as it may not exist on all platforms.

- [x] https://github.com/helix-editor/helix/pull/5692 (merge is +6 vs. +23 of original PR)
- [x] https://github.com/helix-editor/helix/pull/5603 + added functionality of https://github.com/helix-editor/helix/pull/5945

### Suggestions:
* Remove doc comments in editor.rs, they're out of sync with `configuration.md`. Alternatively generate `configugration.md` from them.

## Conflicting PRs (28)

#5411 #5435 #5436 #5491 #5523 #5577 #5608 #5621 #5645 #5660 #5748 #5803 #5839 #5860 #5893 #5931 #6056 #6059 #6061 #6065 #6067 #4307 #4443 #4493 #5199 #2869 #3328 #3393
\* That are mergeable with master and pass all CI checks as of 2023-02-22

### CODE REVIEW

**Comment by gibbz00** [comment]
Also, should the Config field values in editor.rs be removed? They are out of sync with those of configuration.md. Maybe add them back if doc gen functionality is added to them?

File: helix-loader/src/lib.rs [hunk 1]
```diff
@@ -263,11 +263,13 @@ mod merge_toml_tests {
         indent = { tab-width = 4, unit = "    ", test = "aaa" }
         "#;
 
-        let base: Value = toml::from_slice(include_bytes!("../../languages.toml"))
-            .expect("Couldn't parse built-in languages config");
+        // NOTE: Exact duplicate of helix_core::LanguageConfigurations::default()
+        let default: Value =
+            toml::from_str(&std::fs::read_to_string(crate::repo_paths::default_lang_configs()).unwrap())
```
Comment (Proful) [inline helix-loader/src/lib.rs:268]: any case where unwrap will panic?

Comment (gibbz00) [reply helix-loader/src/lib.rs:268]: If the repo languages.toml is moved from root without updating the path in helix_loader::repo_paths!

**Comment by archseer** [comment]
I'd prefer PRs didn't bundle other PRs. While #5411 will get merged I'd rather review it as a standalone unit since it's much more manageable to land one by one.

**Comment by gibbz00** [comment]
How should I treat PRs such as: https://github.com/helix-editor/helix/pull/5839?

It heavily conflicts with this one. Merge it now or only when/if the other is merged?

**Comment by flokli** [comment]
@gibbz00 with https://github.com/helix-editor/helix/pull/5411 merged, can you rebase, and possibly clean up the history a bit? This should make it easier to review.

**Comment by gibbz00** [comment]
> @gibbz00 with #5411 merged, can you rebase, and possibly clean up the history a bit? This should make it easier to review.

I have no idea what the status is on #5635, which this PR builds upon. And given how my other PRs / discussions [have](https://github.com/helix-editor/helix/pull/6067). [been](https://github.com/helix-editor/helix/pull/6056#issuecomment-1437242551). [received](https://github.com/helix-editor/helix/pull/6088#issuecomment-1443854514). , I've moved my focus to other places.

**Comment by pascalkuthe** [comment]
Closing as stale

**Pull request closed without merging**